### PR TITLE
Wrap `wasm-opt` related functions in struct

### PR DESCRIPTION
**Pull request by HCastano**
This PR adds a `WasmOptHandler` struct, which wraps all the functions related to
`wasm-opt`. This lets us ensure that our `wasm-opt` binary is there and valid before we
even try and optimize a binary.

It also makes it easier to pull out certain information about the `wasm-opt` instance
we're using (such as the `wasm-opt` version, which I need for #680).

### CODE REVIEW

File: src/cmd/build.rs [hunk 1]
```diff
@@ -250,6 +249,222 @@ impl CheckCommand {
...
+const WASM_OPT_INSTALLATION_SUGGESTION: &str =
+                "wasm-opt not found! Make sure the binary is in your PATH environment.\n\n\
+                We use this tool to optimize the size of your contract's Wasm binary.\n\n\
+                wasm-opt is part of the binaryen package. You can find detailed\n\
+                installation instructions on https://github.com/WebAssembly/binaryen#tools.\n\n\
+                There are ready-to-install packages for many platforms:\n\
+                * Debian/Ubuntu: apt-get install binaryen\n\
+                * Homebrew: brew install binaryen\n\
+                * Arch Linux: pacman -S binaryen\n\
+                * Windows: binary releases at https://github.com/WebAssembly/binaryen/releases";
```
Comment (cmichi) [inline src/cmd/build.rs:261]: The indentation seems to mismatch the other indentations in the file.

File: src/cmd/build.rs [hunk 2]
```diff
@@ -250,6 +249,222 @@ impl CheckCommand {
...
+/// A helpful struct for interacting with Binaryen's `wasm-opt` tool.
+struct WasmOptHandler {
+    /// The path to the `wasm-opt` binary.
+    wasm_opt_path: PathBuf,
+    /// The optimization level that should be used when optimizing the Wasm binary.
+    optimization_level: OptimizationPasses,
+    /// Whether or not to keep debugging information in the final Wasm binary.
+    keep_debug_symbols: bool,
+    /// The version number of the `wasm-opt` binary being executed.
+    _version: u32,
```
Comment (cmichi) [inline src/cmd/build.rs:272]: Why prefix with `_`?

**Comment by cmichi** [comment]
I think you should put `WasmOptHandler` in its own file, that's just the next consequence of abstracting away from what is happening in that handler and it will make the `build.rs` file less cluttered.

File: src/cmd/build.rs (hunk 2 above)
Comment (HCastano) [reply src/cmd/build.rs:272]: It's not used in this PR, but will be used in #680

**Review by cmichi — APPROVED** [review]
👍

File: src/wasm_opt.rs [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+        let mut command = Command::new(self.wasm_opt_path.as_path());
+        command
+        .arg(dest_wasm.as_os_str())
+        .arg(format!("-O{}", self.optimization_level))
+        .arg("-o")
+        .arg(dest_optimized.as_os_str())
+        // the memory in our module is imported, `wasm-opt` needs to be told that
+        // the memory is initialized to zeroes, otherwise it won't run the
+        // memory-packing pre-pass.
+        .arg("--zero-filled-memory");
```
Comment (ascjones) [inline src/wasm_opt.rs:115]: Surely this whole block should be indented?

File: src/wasm_opt.rs [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+        })?
+        .as_str()
+        .parse()
+        .map_err(|err| {
+            anyhow::anyhow!(
+                "Parsing version number failed with '{:?}' for '{:?}'",
+                err,
+                version_stdout
+            )
+        })?;
```
Comment (ascjones) [inline src/wasm_opt.rs:233]: Same here.

**Review by ascjones — APPROVED** [review]
LGTM, `rustfmt` still on a bender though

File: src/wasm_opt.rs (hunk 3 above)
Comment (ascjones) [inline src/wasm_opt.rs:115]: Surely this whole block should be indented?

File: src/wasm_opt.rs (hunk 4 above)
Comment (ascjones) [inline src/wasm_opt.rs:233]: Should also be indented

File: src/wasm_opt.rs (hunk 3 above)
Comment (HCastano) [reply src/wasm_opt.rs]: RustFmt seems indifferent, but yeah I've indented it

**Pull request merged**